common/cache/cache.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cache
+
+import (
+	"time"
+)
+
+// A Cache is a generalized interface to a cache.  See cache.LRU for a specific
+// implementation (bounded cache with LRU eviction)
+type Cache interface {
+	// Get retrieves an element based on a key, returning nil if the element
+	// does not exist
+	Get(key string) interface{}
+
+	// Put adds an element to the cache, returning the previous element
+	Put(key string, value interface{}) interface{}
+
+	// PutIfNotExist puts a value associated with a given key if it does not exist
+	PutIfNotExist(key string, value interface{}) (interface{}, error)
+
+	// Delete deletes an element in the cache
+	Delete(key string)
+
+	// Release decrements the ref count of a pinned element. If the ref count
```

Comment:
Which function pins the element. I couldn't find any reference in the docs.

Reply:
When you create the LRU cache, you supply optional Options. One of the option is to enable pin. When pin is enabled for a cache, all get request is a pin (with a refcount increase).